File: internal/dto/articles.go
package dto

import (
	"time"

	"github.com/davidmovas/postulator/internal/domain/entities"
)

type Article struct {
	ID            int64   `json:"id"`
	SiteID        int64   `json:"siteId"`
	JobID         *int64  `json:"jobId"`
	TopicID       *int64  `json:"topicId"`
	Title         string  `json:"title"`
	OriginalTitle string  `json:"originalTitle"`
	Content       string  `json:"content"`
	Excerpt       *string `json:"excerpt"`
	WPPostID      int     `json:"wpPostId"`
	WPPostURL     string  `json:"wpPostUrl"`
	WPCategoryIDs []int   `json:"wpCategoryIds"`
	WPTagIDs      []int   `json:"wpTagIds"`
	Status        string  `json:"status"`
	WordCount     *int    `json:"wordCount"`
	Source        string  `json:"source"`
	IsEdited      bool    `json:"isEdited"`
	CreatedAt     string  `json:"createdAt"`
	PublishedAt   *string `json:"publishedAt"`
	UpdatedAt     string  `json:"updatedAt"`
	LastSyncedAt  *string `json:"lastSyncedAt"`

	// SEO & WordPress fields
	Slug             *string `json:"slug"`
	FeaturedMediaID  *int    `json:"featuredMediaId"`
	FeaturedMediaURL *string `json:"featuredMediaUrl"`
	MetaDescription  *string `json:"metaDescription"`
	Author           *int    `json:"author"`
}

func NewArticle(entity *entities.Article) *Article {
	a := &Article{}
	return a.FromEntity(entity)
}

func (d *Article) ToEntity() (*entities.Article, error) {
	createdAt, err := StringToTime(d.CreatedAt)
	if err != nil {
		return nil, err
	}

	updatedAt, err := StringToTime(d.UpdatedAt)
	if err != nil {
		return nil, err
	}

	var publishedAt, lastSyncedAt *time.Time
	if d.PublishedAt != nil {
		var publishedAtTime time.Time
		publishedAtTime, err = StringToTime(*d.PublishedAt)
		if err != nil {
			return nil, err
		}
		publishedAt = &publishedAtTime
	}

	if d.LastSyncedAt != nil {
		var lastSyncedAtTime time.Time
		lastSyncedAtTime, err = StringToTime(*d.LastSyncedAt)
		if err != nil {
			return nil, err
		}
		lastSyncedAt = &lastSyncedAtTime
	}

	return &entities.Article{
		ID:               d.ID,
		SiteID:           d.SiteID,
		JobID:            d.JobID,
		TopicID:          d.TopicID,
		Title:            d.Title,
		OriginalTitle:    d.OriginalTitle,
		Content:          d.Content,
		Excerpt:          d.Excerpt,
		WPPostID:         d.WPPostID,
		WPPostURL:        d.WPPostURL,
		WPCategoryIDs:    d.WPCategoryIDs,
		WPTagIDs:         d.WPTagIDs,
		Status:           entities.ArticleStatus(d.Status),
		WordCount:        d.WordCount,
		Source:           entities.Source(d.Source),
		IsEdited:         d.IsEdited,
		CreatedAt:        createdAt,
		PublishedAt:      publishedAt,
		UpdatedAt:        updatedAt,
		LastSyncedAt:     lastSyncedAt,
		Slug:             d.Slug,
		FeaturedMediaID:  d.FeaturedMediaID,
		FeaturedMediaURL: d.FeaturedMediaURL,
		MetaDescription:  d.MetaDescription,
		Author:           d.Author,
	}, nil
}

func (d *Article) FromEntity(entity *entities.Article) *Article {
	d.ID = entity.ID
	d.SiteID = entity.SiteID
	d.JobID = entity.JobID
	d.TopicID = entity.TopicID
	d.Title = entity.Title
	d.OriginalTitle = entity.OriginalTitle
	d.Content = entity.Content
	d.Excerpt = entity.Excerpt
	d.WPPostID = entity.WPPostID
	d.WPPostURL = entity.WPPostURL
	d.WPCategoryIDs = entity.WPCategoryIDs
	d.WPTagIDs = entity.WPTagIDs
	d.Status = string(entity.Status)
	d.WordCount = entity.WordCount
	d.Source = string(entity.Source)
	d.IsEdited = entity.IsEdited
	d.CreatedAt = TimeToString(entity.CreatedAt)
	d.UpdatedAt = TimeToString(entity.UpdatedAt)

	// SEO & WordPress fields
	d.Slug = entity.Slug
	d.FeaturedMediaID = entity.FeaturedMediaID
	d.FeaturedMediaURL = entity.FeaturedMediaURL
	d.MetaDescription = entity.MetaDescription
	d.Author = entity.Author

	if entity.PublishedAt != nil {
		publishedAt := TimeToString(*entity.PublishedAt)
		d.PublishedAt = &publishedAt
	} else {
		d.PublishedAt = nil
	}

	if entity.LastSyncedAt != nil {
		lastSyncedAt := TimeToString(*entity.LastSyncedAt)
		d.LastSyncedAt = &lastSyncedAt
	} else {
		d.LastSyncedAt = nil
	}

	return d
}

type ArticleListFilter struct {
	SiteID     int64   `json:"siteId"`
	Status     *string `json:"status"`
	Source     *string `json:"source"`
	CategoryID *int    `json:"categoryId"`
	Search     *string `json:"search"`
	SortBy     string  `json:"sortBy"`
	SortOrder  string  `json:"sortOrder"`
	Limit      int     `json:"limit"`
	Offset     int     `json:"offset"`
}

type ArticleListResult struct {
	Articles []*Article `json:"articles"`
	Total    int        `json:"total"`
}

type WPInfoUpdate struct {
	ID          int64   `json:"id"`
	WPPostID    int     `json:"wpPostId"`
	WPPostURL   string  `json:"wpPostUrl"`
	Status      string  `json:"status"`
	PublishedAt *string `json:"publishedAt"`
}

func NewWPInfoUpdate(entity *entities.WPInfoUpdate) *WPInfoUpdate {
	w := &WPInfoUpdate{}
	return w.FromEntity(entity)
}

func (d *WPInfoUpdate) ToEntity() (*entities.WPInfoUpdate, error) {
	var publishedAt *time.Time
	if d.PublishedAt != nil {
		publishedAtTime, err := StringToTime(*d.PublishedAt)
		if err != nil {
			return nil, err
		}
		publishedAt = &publishedAtTime
	}

	return &entities.WPInfoUpdate{
		ID:          d.ID,
		WPPostID:    d.WPPostID,
		WPPostURL:   d.WPPostURL,
		Status:      entities.Status(d.Status),
		PublishedAt: publishedAt,
	}, nil
}

func (d *WPInfoUpdate) FromEntity(entity *entities.WPInfoUpdate) *WPInfoUpdate {
	d.ID = entity.ID
	d.WPPostID = entity.WPPostID
	d.WPPostURL = entity.WPPostURL
	d.Status = string(entity.Status)

	if entity.PublishedAt != nil {
		publishedAt := TimeToString(*entity.PublishedAt)
		d.PublishedAt = &publishedAt
	} else {
		d.PublishedAt = nil
	}

	return d
}

File: frontend/src/models/articles.ts
import { dto } from "@/wailsjs/wailsjs/go/models";

export type ArticleStatus = 'draft' | 'published' | 'pending' | 'private' | 'unknown';
export type ArticleSource = 'generated' | 'manual' | 'imported';

export interface Article {
    id: number;
    siteId: number;
    jobId?: number;
    topicId?: number;
    title: string;
    originalTitle: string;
    content: string;
    excerpt?: string;
    wpPostId: number;
    wpPostUrl: string;
    wpCategoryIds: number[];
    wpTagIds: number[];
    status: ArticleStatus;
    wordCount?: number;
    source: ArticleSource;
    isEdited: boolean;
    createdAt: string;
    publishedAt?: string;
    updatedAt: string;
    lastSyncedAt?: string;

    // SEO & WordPress fields
    slug?: string;
    featuredMediaId?: number;
    featuredMediaUrl?: string;
    metaDescription?: string;
    author?: number;
}

export interface ArticleListFilter {
    siteId: number;
    status?: string;
    source?: string;
    categoryId?: number;
    search?: string;
    sortBy?: string;
    sortOrder?: 'asc' | 'desc';
    limit: number;
    offset: number;
}

export interface ArticleListResult {
    articles: Article[];
    total: number;
}

export interface ArticleCreateInput {
    siteId: number;
    topicId?: number;
    title: string;
    content: string;
    excerpt?: string;
    wpCategoryIds?: number[];
    wpTagIds?: number[];
    slug?: string;
    metaDescription?: string;
}

export interface ArticleUpdateInput {
    id: number;
    title?: string;
    content?: string;
    excerpt?: string;
    wpCategoryIds?: number[];
    wpTagIds?: number[];
    status?: ArticleStatus;
    slug?: string;
    metaDescription?: string;
    featuredMediaId?: number | null;  // null means explicitly clear
    featuredMediaUrl?: string | null; // null means explicitly clear
}

export function mapArticle(x: dto.Article): Article {
    return {
        id: x.id,
        siteId: x.siteId,
        jobId: x.jobId ?? undefined,
        topicId: x.topicId ?? undefined,
        title: x.title,
        originalTitle: x.originalTitle,
        content: x.content,
        excerpt: x.excerpt ?? undefined,
        wpPostId: x.wpPostId,
        wpPostUrl: x.wpPostUrl,
        wpCategoryIds: x.wpCategoryIds || [],
        wpTagIds: x.wpTagIds || [],
        status: x.status as ArticleStatus,
        wordCount: x.wordCount ?? undefined,
        source: x.source as ArticleSource,
        isEdited: x.isEdited,
        createdAt: x.createdAt,
        publishedAt: x.publishedAt ?? undefined,
        updatedAt: x.updatedAt,
        lastSyncedAt: x.lastSyncedAt ?? undefined,
        slug: x.slug ?? undefined,
        featuredMediaId: x.featuredMediaId ?? undefined,
        featuredMediaUrl: x.featuredMediaUrl ?? undefined,
        metaDescription: x.metaDescription ?? undefined,
        author: x.author ?? undefined,
    };
}

export function mapArticleListResult(x: dto.ArticleListResult): ArticleListResult {
    return {
        articles: (x.articles || []).map(mapArticle),
        total: x.total,
    };
}

export const articleStatusLabels: Record<ArticleStatus, string> = {
    draft: 'Draft',
    published: 'Published',
    pending: 'Pending',
    private: 'Private',
    unknown: 'Unknown',
};

export const articleSourceLabels: Record<ArticleSource, string> = {
    generated: 'Generated',
    manual: 'Manual',
    imported: 'Imported',
};

export const articleStatusColors: Record<ArticleStatus, string> = {
    draft: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/30 dark:text-yellow-400',
    published: 'bg-green-100 text-green-800 dark:bg-green-900/30 dark:text-green-400',
    pending: 'bg-blue-100 text-blue-800 dark:bg-blue-900/30 dark:text-blue-400',
    private: 'bg-purple-100 text-purple-800 dark:bg-purple-900/30 dark:text-purple-400',
    unknown: 'bg-gray-100 text-gray-800 dark:bg-gray-900/30 dark:text-gray-400',
};

export interface GenerateContentInput {
    siteId: number;
    providerId: number;
    promptId: number;
    topicId?: number;
    customTopicTitle?: string;
    placeholderValues: Record<string, string>;
}

export interface GenerateContentResult {
    title: string;
    content: string;
    excerpt: string;
    metaDescription: string;
    topicId?: number;
}

export function mapGenerateContentResult(x: dto.GenerateContentResult): GenerateContentResult {
    return {
        title: x.title,
        content: x.content,
        excerpt: x.excerpt,
        metaDescription: x.metaDescription,
        topicId: x.topicId ?? undefined,
    };
}
